polygeist/lib/Dialect/Polygeist/Transforms/ArgumentPromotion.cpp

```diff
@@ -0,0 +1,385 @@
+
+  SmallVector<Value> newCallOperands;
+  newCallOperands.reserve(numCallOperands - numKeys + numMembersPeeled);
+
+  for (unsigned pos = 0; pos < numCallOperands; ++pos) {
+    if (operandToMembersPeeled.find(pos) == operandToMembersPeeled.end()) {
+      newCallOperands.push_back(callOp->getOperand(pos));
+      continue;
+    }
+
+    for (Value peeledMember : operandToMembersPeeled[pos])
+      newCallOperands.push_back(peeledMember);
+  }
+  assert(newCallOperands.size() > numCallOperands);
+
+  callOp->setOperands(newCallOperands);
+  LLVM_DEBUG(llvm::dbgs() << "New call:\n" << callOp << "\n\n");
+}
+
+void Candidate::modifyCallee() {
+
+  auto callable = cast<CallableOpInterface>(callOp.resolveCallable());
+  auto funcOp = cast<func::FuncOp>(callable.getCallableRegion()->getParentOp());
+  Region &callableRgn = funcOp.getFunctionBody();
+  ArrayAttr origArgAttrs = funcOp.getAllArgAttrs();
+  SmallVector<Attribute> newArgAttrs;
+
+  // Replace the old arguments with the new ones.
+  unsigned pos = 0;
+  const unsigned orinNumArgs = callableRgn.getNumArguments();
+  for (unsigned origPos = 0; origPos < orinNumArgs; ++origPos) {
+    if (operandToMembersPeeled.find(origPos) == operandToMembersPeeled.end()) {
+      newArgAttrs.push_back(origArgAttrs ? origArgAttrs[origPos] : Attribute());
+      continue;
+    }
+
+    // Replace the argument at 'pos' with the new arguments we peeled.
+    const SmallVector<Value> &newArgs = operandToMembersPeeled[origPos];
+    replaceArgumentWith(pos, callableRgn, newArgs, newArgAttrs);
+
+    // Delete the original argument.
+    pos += newArgs.size();
+    callableRgn.eraseArgument(pos);
+  }
+
+  auto newFuncType =
+      FunctionType::get(funcOp.getContext(), callableRgn.getArgumentTypes(),
+                        funcOp.getResultTypes());
+
+  funcOp.setFunctionType(newFuncType);
+  funcOp.setAllArgAttrs(newArgAttrs);
+
+  LLVM_DEBUG(llvm::dbgs() << "New Callee:\n" << funcOp << "\n\n");
+}
+
+void Candidate::replaceArgumentWith(unsigned pos, Region &callableRgn,
+                                    const SmallVector<Value> &newArgs,
+                                    SmallVector<Attribute> &newArgAttrs) const {
+  assert(!newArgs.empty() && "Expecting a non-empty vector");
+
+  Value origArg = callableRgn.getArgument(pos);
+
+  for (unsigned offset = 0; offset < newArgs.size(); ++offset) {
+    callableRgn.insertArgument(pos + offset, newArgs[offset].getType(),
+                               callableRgn.getLoc());
+    // TODO: add attribute NoAlias on the new argument.
+    newArgAttrs.push_back(Attribute());
+  }
+
+  // Replace uses of the original argument with the new arguments injected.
+  replaceUsesOfArgument(origArg, pos, callableRgn);
+}
+
+void Candidate::replaceUsesOfArgument(Value origArg, unsigned pos,
+                                      Region &callableRgn) const {
+  for (OpOperand &use : origArg.getUses()) {
+    assert(isa<polygeist::SubIndexOp>(use.getOwner()) &&
```

Comment:
instead of just assertion, should we check in advance as one of the candidate requirement.